File: src/main/java/com/xpdustry/nucleus/network/InetAddressWhitelist.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.network;

import com.xpdustry.nucleus.database.PostgresDatabase;
import java.net.InetAddress;

public final class InetAddressWhitelist {

    private final PostgresDatabase database;

    public InetAddressWhitelist(final PostgresDatabase database) {
        this.database = database;
    }

    public boolean contains(final InetAddress address) {
        return this.database.withTransaction(handle -> handle.prepareStatement("""
                        SELECT 1 FROM "address_whitelist" "a"
                        WHERE ?::inet <<= "a"."address" LIMIT 1
                        """)
                .bind(address.getHostAddress())
                .executeSingleSelect(_ -> Boolean.TRUE)
                .orElse(false));
    }

    public void insert(final InetAddress address, final String reason) {
        this.database.withTransaction(handle -> handle.prepareStatement("""
                        INSERT INTO "address_whitelist"("address", "reason") VALUES (?, ?)
                        """)
                .bind(address.getHostAddress())
                .bind(reason)
                .executeSingleUpdate());
    }
}

File: src/main/kotlin/com/xpdustry/nucleus/network/InetAddressWhitelist.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.network

import com.xpdustry.nucleus.database.PostgresDatabase
import java.net.InetAddress
import kotlin.String

class InetAddressWhitelist(private val database: PostgresDatabase) {
    suspend fun contains(address: InetAddress): Boolean {
        return this.database.transaction {
            """
                SELECT 1 FROM "address_whitelist" "a"
                WHERE ?::inet <<= "a"."address" LIMIT 1
                """
                .asPreparedStatement()
                .push(address.hostAddress)
                .executeSingleSelect { true }
        }
    }

    suspend fun insert(address: InetAddress, reason: String) {
        this.database.transaction {
            """
            INSERT INTO "address_whitelist"("address", "reason") VALUES (?, ?)
            """
                .asPreparedStatement()
                .push(address.hostAddress)
                .push(reason)
                .executeSingleUpdate()
        }
    }
}
